Classes/NodeTemplateDumper/NodeTemplateDumper.php

```diff
@@ -0,0 +1,258 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flowpack\NodeTemplates\NodeTemplateDumper;
+
+use Neos\ContentRepository\Domain\Model\ArrayPropertyCollection;
+use Neos\ContentRepository\Domain\Model\Node;
+use Neos\ContentRepository\Domain\Model\NodeInterface;
+use Neos\ContentRepository\Domain\Projection\Content\TraversableNodeInterface;
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\I18n\EelHelper\TranslationHelper;
+use Symfony\Component\Yaml\Yaml;
+
+/** @Flow\Scope("singleton") */
+class NodeTemplateDumper
+{
+    /**
+     * @var TranslationHelper
+     * @Flow\Inject
+     */
+    protected $translationHelper;
+
+    /**
+     * Dump the node tree structure into a NodeTemplate Yaml structure.
+     * References to Nodes and non-primitive property values are commented out in the Yaml.
+     *
+     * @param Node|NodeInterface|TraversableNodeInterface $startingNode specified root node of the node tree to dump
+     * @return string yaml representation of the node template
+     */
+    public function createNodeTemplateYamlDumpFromSubtree(Node $startingNode): string
+    {
+        $comments = Comments::empty();
+
+        $nodeType = $startingNode->getNodeType();
+        if ($nodeType->isOfType('Neos.Neos:Document')) {
+            $template = $this->nodeTemplateFromDocumentNodes([$startingNode], $comments);
+        } elseif ($nodeType->isOfType('Neos.Neos:Content') || $nodeType->isOfType('Neos.Neos:ContentCollection')) {
+            $template = $this->nodeTemplateFromContentNodes([$startingNode], $comments);
+        } else {
+            throw new \InvalidArgumentException("Node {$startingNode->getIdentifier()} must be one of Neos.Neos:Document,Neos.Neos:Content,Neos.Neos:ContentCollection.");
+        }
+
+        foreach ($template as $firstEntry) {
+            break;
+        }
+        assert(isset($firstEntry));
+
+        $templateRoot = [
+            "template" => array_filter([
+                "properties" => $firstEntry["properties"] ?? null,
+                "childNodes" => $firstEntry["childNodes"] ?? null,
+            ])
+        ];
+
+        $yaml = Yaml::dump($templateRoot, 99, 2, Yaml::DUMP_EMPTY_ARRAY_AS_SEQUENCE | Yaml::DUMP_NULL_AS_TILDE);
+
+        $yamlWithComments = $comments->renderCommentsInYamlDump($yaml);
+
+        return $yamlWithComments;
+    }
+
+    /** @param array<Node|NodeInterface> $nodes */
+    private function nodeTemplateFromDocumentNodes(array $nodes, Comments $comments): array
```

Comment:
The `from` methods should be merged by adding a simple `isDocumentNode` condition where needed